### PR TITLE
Derive `Ord`, `PartialOrd` and `Hash` for `SocketAddr*`

### LINKED ISSUE #116711: SocketAddrV6 Ord and Eq disagree with each other
I tried this code:

https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=5c3bc19334a7cf61d10517e874f8cf0b

I expected to see this happen: `Equals, true`

Instead, this happened: `Equals, false`

### Meta
yeah so SocketAddr behaves differently in a HashMap vs BTreeMap. Not only is this surprising but it also conflicts with the documentation for Ord & PartialOrd.

### PR DESCRIPTION
**Pull request by WaffleLapkin**
Fixes #116711

The main pain of this PR is to fix the buggy impl of `Ord` for `SocketAddrV6`, which ignored half of the fields (while `PartialEq` is derived):
https://github.com/rust-lang/rust/blob/4603f0b8afb495ae56cd4c8f70d5d478d906ac54/library/core/src/net/socket_addr.rs#L99-L106

https://github.com/rust-lang/rust/blob/4603f0b8afb495ae56cd4c8f70d5d478d906ac54/library/core/src/net/socket_addr.rs#L676

For me it looks like a simple copy-paste error made in https://github.com/rust-lang/rust/pull/72239 (copy from v4 impl) (cc @hch12907), as I don't see this behavior being mentioned anywhere on the PR and it also does not respect `cmp` trait "rules". I also do not see any reasons for those impls to _not_ be derived.

It's a shame we did not notice this for 28 versions/3 years. I guess this is a bug fix, but I'm not sure what the process here should be.

r? libs

### CODE REVIEW

**Comment by workingjubilee** [comment]
Note that this does not change whether there is an impl insofar as I can see, but they do change the implementation's behavior.

**Comment by WaffleLapkin** [comment]
Yes, the only functional change is `SocketAddrV6::{cmp,partial_cmp}`  also considering `flowinfo` and `scope_id` fields. No impls are removed.p

**Comment by hch12907** [comment]
As the original author of #72239 - this change makes sense.

But I have a question: does it make sense to include `flowinfo`? Unlike the other fields which are fairly fixed, the flow label is "a high-entropy identifier of a flow of packets between a source and destination" (according to [wiki](https://en.wikipedia.org/wiki/IPv6_packet#Fixed_header)).

>I also do not see any reasons for those impls to not be derived.

Note that `SocketAddrV6` was originally just a wrapper of `sockaddr_in6` from libc. From memory, `sockaddr_in6` did not implement `(Partial)Ord` which was why a simple `#[derive]` was not possible.

File: library/core/src/net/socket_addr.rs [hunk 1]
```diff
@@ -96,7 +94,7 @@ pub struct SocketAddrV4 {
 /// assert_eq!(socket.ip(), &Ipv6Addr::new(0x2001, 0xdb8, 0, 0, 0, 0, 0, 1));
 /// assert_eq!(socket.port(), 8080);
 /// ```
-#[derive(Copy, Clone, Eq, PartialEq)]
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
```
Comment (hch12907) [inline library/core/src/net/socket_addr.rs:97]: Can you add some tests for `SocketAddrV6`s with different scope IDs and flow info?

**Comment by workingjubilee** [comment]
>   A flow label is assigned to a flow by the flow's source node.  New
   flow labels must be chosen (pseudo-)randomly and uniformly from the
   range 1 to FFFFF hex.  The purpose of the random allocation is to
   make any set of bits within the Flow Label field suitable for use as
   a hash key by routers, for looking up the state associated with the
   flow.
>
>   All packets belonging to the same flow must be sent with the same
   source address, destination address, and flow label.

So... I think probably yes?

**Comment by hch12907** [comment]
Looks good to me.

File: library/core/src/net/socket_addr.rs (hunk 1 above)
Comment (WaffleLapkin) [reply library/core/src/net/socket_addr.rs:97]: Does [5c13c69](https://github.com/rust-lang/rust/pull/116714/commits/5c13c69f63dccdc637d28b403ae1fb687e0a5143) look good to you?

Comment (hch12907) [reply library/core/src/net/socket_addr.rs:97]: Yep, thanks!

**Comment by joshtriplett** [comment]
Merging per review in libs-api meeting.

**Pull request merged**